daemon/policy.go

```diff
@@ -77,46 +76,65 @@ func (d *Daemon) TriggerPolicyUpdates(added []policy.NumericIdentity) *sync.Wait
 		// FIXME: Invalidate only cache that is affected
 		d.invalidateCache()
 	}
-
-	d.GetPolicyRepository().Mutex.RLock()
-	d.EnablePolicyEnforcement()
-	d.GetPolicyRepository().Mutex.RUnlock()
 	return endpointmanager.TriggerPolicyUpdates(d)
 }
 
 // UpdateEndpointPolicyEnforcement returns whether policy enforcement needs to be
 // enabled for the specified endpoint.
 //
-// Must be called with e.Consumable.Mutex and d.GetPolicyRepository().Mutex held
-func (d *Daemon) UpdateEndpointPolicyEnforcement(e *endpoint.Endpoint) bool {
-	if d.EnablePolicyEnforcement() {
+// Must be called with e.Consumable.Mutex and d.GetPolicyRepository().Mutex held.
+func (d *Daemon) EnableEndpointPolicyEnforcement(e *endpoint.Endpoint) bool {
+	// First check if policy enforcement should be enabled at the daemon level.
+	// If policy enforcement is enabled for the daemon, then it has to be
+	// enabled for the endpoint.
+
+	config.EnablePolicyMU.Lock()
+	daemonPolicyEnable, _ := d.EnablePolicyEnforcement()
+	defer config.EnablePolicyMU.Unlock()
+	if daemonPolicyEnable {
 		return true
 	} else if d.conf.EnablePolicy == endpoint.DefaultEnforcement && d.conf.IsK8sEnabled() {
-		// Check if rules match the labels for this endpoint.
-		// If so, enable policy enforcement.
+		// Default mode + K8s means that if rules contain labels that match
+		// this endpoint, then enable policy enforcement for this endpoint.
 		return d.GetPolicyRepository().GetRulesMatching(e.Consumable.LabelArray)
 	}
+	// If policy enforcement isn't enabled for the daemon, or we are not running
+	// in "default" mode in tandem with K8s, we do not enable policy enforcement
+	// for the endpoint.
+	// This means one of the following:
+	// * daemon policy enforcement mode is 'never', so no policy enforcement
+	//   should be applied to the specified endpoint.
+	// * if we are not running K8s and are running in 'default' mode, we do not
+	//   enable policy enforcement on a per-endpoint basis (i.e., outside of the
+	//   scope of this function).
 	return false
 }
 
 // EnablePolicyEnforcement returns whether policy enforcement needs to be
-// enabled for the daemon.
+// enabled at the daemon-level.
```

Comment:
This is in one of the earlier commits in the branch - I have reverted this functionality in a later commit. I'm going to re-order the commits in the branch, and will re-push once that's done